### PR TITLE
 nixos/nix-channel: only try to remove the nix-channel binary if it exists

### PR DESCRIPTION
**Pull request by r-vdp**
###### Description of changes

~~Since #242098 we remove the `nix-channel` binary from the system path if the nix channel is disabled. This binary is not necessarily present though, for instance when `system.disableInstallerTools` is true. So we introduce a test to check whether it exists before removing it.~~

EDIT: my analysis was wrong, I think @xaverdh below is correct. The fix remains the same though.

Ping: @roberth 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by R-VdP** [comment]
Also, #238557 introduces a simple switch to disable individual commands, which would be a cleaner way to implement this.

File: nixos/modules/config/nix-channel.nix [hunk 1]
```diff
@@ -86,7 +86,9 @@ in
       '';
 
     environment.extraSetup = mkIf (!cfg.channel.enable) ''
-      rm $out/bin/nix-channel
+      if [ -f "$out/bin/nix-channel" ]; then
+        rm $out/bin/nix-channel
+      fi
```
Comment (roberth) [inline nixos/modules/config/nix-channel.nix:91]:
```suggestion
      rm -f $out/bin/nix-channel
```

**Comment by roberth** [comment]
> if the nix channel is disabled.

I think you mean `nix-channel`, not a channel.



> This binary is not necessarily present though, for instance when `system.disableInstallerTools` is true.

I've looked at the code and I see no reason why a normal nix package wouldn't be present in `environment.systemPackages` if that option is set. Also the removal code is only present when `config.nix.enable == true`.
Are you sure this is the mechanism that leads to such an error?

**Comment by xaverdh** [comment]
> 
> > This binary is not necessarily present though, for instance when `system.disableInstallerTools` is true.
> 
> I've looked at the code and I see no reason why a normal nix package wouldn't be present in `environment.systemPackages` if that option is set. Also the removal code is only present when `config.nix.enable == true`. Are you sure this is the mechanism that leads to such an error?

Indeed, when I tested this pr, the error came from the fact that the user environment does not contain anything from `nix`, since that is installed system wide.

**Comment by xaverdh** [comment]
So I think the issue is that https://github.com/NixOS/nixpkgs/blob/d75ba3d7a2b5c4952f89f498e435cb5648c0f0cb/nixos/modules/config/users-groups.nix#L678C55-L678C64 references `postBuild` from https://github.com/NixOS/nixpkgs/blob/d75ba3d7a2b5c4952f89f498e435cb5648c0f0cb/nixos/modules/config/system-path.nix#L176, so it also inherits the `rm [...]` part

**Comment by R-VdP** [comment]
I think you're right. The fix is the same though, so it would still be good to merge this.

**Review by xaverdh — APPROVED** [review]

**Comment by roberth** [comment]
> `users-groups.nix`

Yikes.

> Shell fragments to be run after the system environment has been created. This should only be used for things that need to modify the internals of the environment, e.g. generating MIME caches. The environment being built can be accessed at $out.

That documentation is out of date.
 - https://github.com/NixOS/nixpkgs/pull/31283#issuecomment-1649455143

**Comment by priegger** [comment]
@roberth : Since #242098 was backported to 23.05, this one should be backported, too, I think.

Can this be done automatically or should I create a backport pr manually?

**Pull request merged**